subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentVariantIdentifier.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.artifacts;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+
+/**
+ * Identifies a variant of a component by module identifier and variant name.
+ *
+ * @since 6.0
+ */
+@Incubating
+public interface ComponentVariantIdentifier {
+
+    /**
+     * Returns the component identifier.
+     */
+    <T extends ComponentIdentifier> T getId();
```

Comment:
Added the type parameter for a better Kotlin DSL experience.

Reply:
Might give a better experience, but that's not type safe at all, right?
See other comment on the sample.